pkg/controller/add_cluster.go

```diff
@@ -170,24 +172,61 @@ func (c *Controller) EnsurePlacement(cluster *myspec.M3DBCluster) error {
 	return nil
 }
 
-// EnsureNamespace will retrieve current namespaces to ensure one matches
-// the cluster name or create a new namespace to match the cluster name
-func (c *Controller) EnsureNamespace(cluster *myspec.M3DBCluster) error {
+// EnsureNamespace will retrieve current namespaces to ensure one matches the
+// cluster name or create a new namespace to match the cluster name. Returns a
+// bool indicating whether a namespace was created and any errors encountered.
+func (c *Controller) EnsureNamespace(cluster *myspec.M3DBCluster) (bool, error) {
 	// Get namespace
 	namespaces, err := c.namespaceClient.List()
 	if err != nil {
 		c.logger.Error("failed to get namespace ", zap.Error(err))
-		return err
+		return false, err
 	}
 	for _, md := range namespaces {
 		if md.ID().Equal(ident.StringID(cluster.GetName())) {
 			c.logger.Info("namespace found", zap.String("ns", md.ID().String()))
-			return nil
+			return false, nil
 		}
 	}
 	if err = c.namespaceClient.Create(cluster.GetObjectMeta().GetName()); err != nil {
 		c.logger.Error("failed to create namespace", zap.Error(err))
+		return false, err
+	}
+	return true, nil
+}
+
+func (c *Controller) ensureServices(cluster *myspec.M3DBCluster) error {
+	// TODO(schallert): support updating service spec, not sure if this only
+	// handles creation.
+	m3dbSvc, found := c.getServiceConfig(_M3DBSvcName, cluster.Spec.ServiceConfigurations)
+	if !found {
+		err := errors.New("m3db service not found in spec")
+		c.logger.Error(err.Error())
+		c.recorder.Event(cluster, corev1.EventTypeWarning, "NoServiceConfig", "could not create m3db service, missing service config in spec")
+		return err
+	}
+
+	if err := c.k8sclient.EnsureService(cluster, m3dbSvc); err != nil {
+		err = fmt.Errorf("error creating m3db service: %v", err)
+		c.logger.Error(err.Error())
+		c.recorder.Event(cluster, corev1.EventTypeWarning, "ServiceCreateError", err.Error())
 		return err
 	}
+
+	coordSvc, found := c.getServiceConfig(_M3CoordinatorSvcName, cluster.Spec.ServiceConfigurations)
+	if !found {
+		err := errors.New("coordinator service not found in spec")
+		c.logger.Error(err.Error())
+		c.recorder.Event(cluster, corev1.EventTypeWarning, "NoServiceConfig", "could not create m3coordinator service, missing service config in spec")
+		return err
+	}
+
+	if err := c.k8sclient.EnsureService(cluster, coordSvc); err != nil {
+		err = fmt.Errorf("error creating m3coordinator service: %v", err)
+		c.logger.Error(err.Error())
+		c.recorder.Event(cluster, corev1.EventTypeWarning, "ServiceCreateError", err.Error())
```

Comment:
I like these events, how liberal should we be with them?

Reply:
I definitely want to use them extensively. Even something like creating a statefulset causes 6+ info-level events. Once the behavior is more defined I think we can depend on them more